tests/kokkos-based/helpers.hpp

```diff
@@ -166,60 +195,216 @@ bool is_same_vector(
           typename LayoutPolicy2,
-          typename AccessorPolicy2,
-          typename ToleranceType>
+          typename AccessorPolicy2>
 bool is_same_matrix(
-    mdspan<ElementType, extents<dynamic_extent, dynamic_extent>, LayoutPolicy1, AccessorPolicy1> A,
-    mdspan<ElementType, extents<dynamic_extent, dynamic_extent>, LayoutPolicy2, AccessorPolicy2> B,
-    ToleranceType tolerance)
+    mdspan<ElementType1, extents<dynamic_extent, dynamic_extent>, LayoutPolicy1, AccessorPolicy1> A,
+    mdspan<ElementType2, extents<dynamic_extent, dynamic_extent>, LayoutPolicy2, AccessorPolicy2> B)
 {
   const auto ext0 = A.extent(0);
   const auto ext1 = A.extent(1);
   if (B.extent(0) != ext0 or B.extent(1) != ext1)
     return false;
   const auto A_view = KokkosKernelsSTD::Impl::mdspan_to_view(A);
   const auto B_view = KokkosKernelsSTD::Impl::mdspan_to_view(B);
-  // Note: reducint to `int` because Kokkos can complain on `bool` not being 
+  // Note: reducint to `int` because Kokkos can complain on `bool` not being
   //       aligned with int32 and deny it for parallel_reduce()
   using diff_type = int;
   diff_type is_different = false;
   Kokkos::parallel_reduce(ext0,
     KOKKOS_LAMBDA(std::size_t i, diff_type &diff) {
         for (decltype(i) j = 0; j < ext1; ++j) {
-          const auto d = scalar_diff(A_view(i, j), B_view(i, j));
-          diff = diff || (d > tolerance);
+          const bool d = A_view(i, j) != B_view(i, j);
+          diff = diff || d;
         }
 	    }, Kokkos::LOr<diff_type>(is_different));
   return !is_different;
 }
 
+template <typename ElementType,
+          typename LayoutPolicy1,
+          typename AccessorPolicy1>
+bool is_same_matrix(
+    mdspan<ElementType, extents<dynamic_extent, dynamic_extent>, LayoutPolicy1, AccessorPolicy1> A,
+    const std::vector<ElementType> &B)
+{
+  return is_same_matrix(A, make_mdspan(B.data(), A.extent(0), A.extent(1)));
+}
+
+template <typename ElementType,
+          typename LayoutPolicy1,
+          typename AccessorPolicy1>
+bool is_same_matrix(const std::vector<ElementType> &A,
+    mdspan<ElementType, extents<dynamic_extent, dynamic_extent>, LayoutPolicy1, AccessorPolicy1> B)
+{
+  return is_same_matrix(make_mdspan(A.data(), B.extent(0), B.extent(1)), B);
+}
+
+template <typename ElementType,
+          typename LayoutPolicy1,
+          typename AccessorPolicy1,
+          typename LayoutPolicy2,
+          typename AccessorPolicy2>
+auto matrix_abs_diff(
+    mdspan<ElementType, extents<dynamic_extent, dynamic_extent>, LayoutPolicy1, AccessorPolicy1> A,
+    mdspan<ElementType, extents<dynamic_extent, dynamic_extent>, LayoutPolicy2, AccessorPolicy2> B)
+{
+  const auto ext0 = A.extent(0);
+  const auto ext1 = A.extent(1);
+  using RetType = decltype(scalar_abs_diff(A(0, 0), B(0, 0)));
+  if (B.extent(0) != ext0 or B.extent(1) != ext1) {
+    return std::numeric_limits<RetType>::max(); // very, very different
+  }
+  return vector_abs_diff(
```

Comment:
[REQUIRED]

This is incorrect for any layout other than `layout_left` or `layout_right`.  In particular, this may give incorrect results for any `layout_stride` matrix with leading dimension greater than the number of rows (a use case that the C and Fortran BLAS both support).

There are two approaches for fixing this.

1. Constrain _both_ layouts to be either `layout_left` or `layout_right`.
2. Write a generic implementation.

Reply:
Thank you for pointing this out! Fixed with generic implementation on https://github.com/kokkos/stdBLAS/pull/193/commits/a945906066e35af077cbdcbf0cfdccb55ea2eeb3.